flang/docs/ParallelFortranRuntime.md

```diff
@@ -0,0 +1,1987 @@
+  integer(c_int), allocatable, intent(out) :: failed_images(:)
+end subroutine
+```
+
+### `prif_stopped_images`
+
+**Description**: Determine the image indices of any images known to have initiated
+  normal termination.
+    
+```
+subroutine prif_stopped_images(team, stopped_images)
+  type(prif_team_type), intent(in), optional :: team
+  integer(c_int), allocatable, intent(out) :: stopped_images(:)
+end subroutine
+```
+
+### `prif_image_status`
+
+**Description**: Determine the image execution state of an image
+    
+```
+impure elemental subroutine prif_image_status(image, team, image_status)
+  integer(c_int), intent(in) :: image
+  type(prif_team_type), intent(in), optional :: team
+  integer(c_int), intent(out) :: image_status
+end subroutine
+```
+**Further argument descriptions**:
+
+* **`image`**: the image index of the image in the given or current team for
+  which to return the execution status
+* **`team`**: if provided, the team from which to identify the image
+* **`image_status`**: defined to the value `PRIF_STAT_FAILED_IMAGE` if the identified
+    image has failed, `PRIF_STAT_STOPPED_IMAGE` if the identified image has initiated
+    normal termination, otherwise zero.
+
+## Storage Management
+
+### Common Arguments
+
+* **`coarray_handle`**
+  * Argument for many of the coarray access procedures
+  * scalar of type [`prif_coarray_handle`](#prif_coarray_handle)
+  * is a handle for the descriptor of an established coarray
+* **`cosubscripts`**
+  * Argument for many of the coarray access procedures
+  * 1d assumed-shape array of type `integer(c_intmax_t)`
+  * correspond to the cosubscripts appearing in a _coindexed-named-object_ reference
+* **`value`** or `current_image_buffer`
+  * Argument for `put` and `get` operations
+  * assumed-rank array of `type(*)` or `type(c_ptr)`
+  * It is the value to be sent in a `put` operation, and is assigned the value
+    retrieved in the case of a `get` operation
+* **`image_num`**
+  * an argument identifying the image to be communicated with
+  * is permitted to identify the current image
+  * the image index is always relative to the initial team, unless otherwise specified
+
+### `prif_allocate_coarray`
+
+**Description**: This procedure allocates memory for a coarray. 
+  This call is collective over the current team.  Calls to
+  `prif_allocate_coarray` will be inserted by the compiler when there is an explicit
+  coarray allocation or at the beginning of a program to allocate space for
+  statically declared coarrays in the source code. The PRIF implementation will
+  store the coshape information in order to internally track it during the
+  lifetime of the coarray.
+    
+```
+subroutine prif_allocate_coarray( &
+    lcobounds, ucobounds, lbounds, ubounds, element_length, &
+    final_func, coarray_handle, allocated_memory, &
+    stat, errmsg, errmsg_alloc)
+  integer(kind=c_intmax_t), intent(in) :: lcobounds(:), ucobounds(:)
```

Comment:
From Section 18.3.7 Interoperability of procedures and procedure interfaces
> A Fortran procedure interface is interoperable with a C function prototype if
> ...
> any dummy argument without the VALUE attribute corresponds to a formal parameter of the prototype that is of a pointer type, and either
> ...
> the dummy argument is ..., assumed-shape, ... and the formal parameter is a pointer to CFI_cdesc_t, or

So, it is interoperable and C receives a CFI_cdesc_t.

Reply:
Thank you for your answers,  which helped me clarify some points.

Reply:
Hi  

I'd like to come back to you after your answers and the problems I'm having with the lowering. In LLVM, the lowering of operations to runtime function calls is done in C++. Calling your interface directly  in this state is not possible due to the interoperability problems.  I have several questions :  

-  What is the impact of adding the BIND(C) attribute for all PRIF subroutines in order to solve the interoperability issues with assumed-shape arrays ? Bind C will help. Is the signature of the PRIF subroutines fixed? If the PRIF interface is fixed as it is, it will be necessary to provide to LLVM a wrapper that will call PRIF's subroutines.  Is this an acceptable and correct solution?  

- If not fixed, how can we call some PRIF subroutines with prif_coarray_handle, if this type is opaque to the compiler with the information that we have after the lowering ? Perhaps could you add a subroutine to retrive the correct coarray_handle with some basic informations like the base_addr ?  Same for prif_team_type, how can we lower/create an entity with this type after the lowering pass.  
  
Thanks in advance for your answers

Reply:
Additional descriptions in V0.4 clarify that PRIF is a Fortran interface. We have had a meeting with @JDPailleux to discuss the issues mentioned above. We are hopeful that this is a workable solution.